libopenage/curve/array.h

```diff
@@ -0,0 +1,244 @@
+		}
+
+		void operator++() {
+			this->offset++;
+		}
+
+		bool operator!=(const Array<T, Size>::Iterator &rhs) const {
+			return this->offset != rhs.offset;
+		}
+
+
+	private:
+		size_t offset;
+		Array<T, Size> *curve;
+		time::time_t time;
+	};
+
+
+	Iterator begin(const time::time_t &time = time::TIME_MAX);
+
+	Iterator end(const time::time_t &time = time::TIME_MAX);
+
+
+private:
+	std::array<KeyframeContainer<T>, Size> container;
+
+	//hint for KeyframeContainer operations
+	mutable std::array<size_t, Size> last_element = {};
+
+	/**
+	* Identifier for the container
+	*/
+	const size_t _id;
+
+	/**
+	* Human-readable identifier for the container
+	*/
+	const std::string _idstr;
+
+	/**
+	* The eventloop this curve was registered to
+	*/
+	const std::shared_ptr<event::EventLoop> loop;
+};
+
+
+template <typename T, size_t Size>
+std::pair<time::time_t, T> Array<T, Size>::frame(const time::time_t &t, const size_t index) const {
+	size_t frmae_index = container[index].last(t, this->last_element[index]);
+	this->last_element[index] = frmae_index;
+	return container[index].get(frmae_index).make_pair();
+}
+
+template <typename T, size_t Size>
+std::pair<time::time_t, T> Array<T, Size>::next_frame(const time::time_t &t, const size_t index) const {
+	size_t frmae_index = container[index].last(t, this->last_element[index]);
+	this->last_element[index] = frmae_index;
+	return container[index].get(frmae_index + 1).make_pair();
+}
+
+template <typename T, size_t Size>
+T Array<T, Size>::get(const time::time_t &t, const size_t index) const {
+	return this->frame(t, index).second;
+}
+
+template <typename T, size_t Size>
+std::array<T, Size> Array<T, Size>::get_all(const time::time_t &t) const {
+	return [&]<auto... I>(std::index_sequence<I...>) {
+		return std::array<T, Size>{this->get(t, I)...};
+	}(std::make_index_sequence<Size>{});
+}
+
+template <typename T, size_t Size>
+consteval size_t Array<T, Size>::size() const {
+	return Size;
+}
+
+
+template <typename T, size_t Size>
+void Array<T, Size>::set_insert(const time::time_t &t, const size_t index, T value) {
+	this->last_element[index] = this->container[index].insert_after(Keyframe(t, value), this->last_element[index]);
+}
+
+
+template <typename T, size_t Size>
+void Array<T, Size>::set_last(const time::time_t &t, const size_t index, T value) {
+
+	size_t frame_index = this->container[index].insert_after(Keyframe(t, value), this->last_element[index]);
+	this->last_element[index] = frame_index;
+	this->container[index].erase_after(frame_index);
+}
+
+
+template <typename T, size_t Size>
+void Array<T, Size>::set_replace(const time::time_t &t, const size_t index, T value) {
+	this->container[index].insert_overwrite(Keyframe(t, value), this->last_element[index]);
+}
+
+template <typename T, size_t Size>
+void Array<T, Size>::sync(const Array<T, Size> &other, const time::time_t &start) {
+	for (int i = 0; i < Size; i++) {
+		this->container[i].sync(other.container[i], start);
+	}
```

Comment:
If you are iterating over every element in the container anyway, you should use a for-each loop.

Reply:
other.container[i] still needs to be indexed?